### ISSUE TITLE
Procedural intersections

### ISSUE THREAD

**Issue by TheDuckCow** [TheDuckCow/godot-road-generator#121]
The initial scope of the intersection support is going to be using pre-generated scenes/objects. That's a good first step, but we eventually want to have procedural intersections. This task is to represent this work.

## Proposal plan

See the primary work plan and proposal in [this post](https://github.com/TheDuckCow/godot-road-generator/issues/121#issuecomment-3374038644) further down.

## Progress

This section is being edited by TheDuckCow as progress occurs. The above plan is broken down into the following steps and follow up work:

# Steps
- [x] Create a dedicated feature branch, from which other branches will be created @TheDuckCow
  - For anyone curious about the latest code, see the `feature-121-proc-intersections` branch [here](https://github.com/TheDuckCow/godot-road-generator/tree/feature-121-proc-intersections)
- [x] Create the architecture @Picorims  https://github.com/TheDuckCow/godot-road-generator/pull/298
- [x] Create the NGon simple geometry @Picorims  https://github.com/TheDuckCow/godot-road-generator/pull/298
- In parallel:
    - [ ] Do mesh generation for more sophisticated smoothed nway geo @Picorims  https://github.com/TheDuckCow/godot-road-generator/pull/323
    - [x] add road properties including collisions (@TheDuckCow #307)
    - Do UI/UX interactions @TheDuckCow:
        - [x] Initial RoadIntersection gizmo and container-level refreshing https://github.com/TheDuckCow/godot-road-generator/pull/310
        - [x] Update how RoadPoints connect to intersections, so they don't count as edges https://github.com/TheDuckCow/godot-road-generator/pull/314
        - [x] Translate RoadPoint should refresh intersection geometry https://github.com/TheDuckCow/godot-road-generator/pull/314
        - [x] Update the connection tool to support converting a RoadPoint into an intersection with undo/redo https://github.com/TheDuckCow/godot-road-generator/pull/315
        - [x] Support using the disconnection tool to remove individual arms of an intersection at a time https://github.com/TheDuckCow/godot-road-generator/pull/315
        - [ ] Re-introduce the RoadContainer snapping tools https://github.com/TheDuckCow/godot-road-generator/pull/325
    - [x] Implement NGon mesh UV mapping + extra should loop to draw shoulder: https://github.com/TheDuckCow/godot-road-generator/issues/308 and https://github.com/TheDuckCow/godot-road-generator/pull/313
    - [x] Fix the NGon asymmetric should width https://github.com/TheDuckCow/godot-road-generator/pull/309
    - [ ] Implement RoadLane generation for NGon (workflow needs to be determined)
    - [ ] Implement edge curves for intersections as a general utility so it can also apply to other intersections added
- [ ] Updated the integration with Terrain3D to account for ngons/smoothed nway intersection
- [ ] Support underside geo with both ngon/smoothed nway intersection

**Comment by TheDuckCow** [comment]
While we are not tackling this just yet, I did come across [this video](https://youtu.be/G5Iq-jxgZn0?si=P2ttQrjOWdEgLPVG&t=942) noting that at 15:42 it shows some nice details of the basis of intersection generation:

Transcribing and extrapolating with my understanding:

- They need enough space (lead in) to round the corners
- Need to support the width of the roads entering
- In between, you create "fillets" which create the rounding of the edges between the sides.
- And then fill with some geometry in the middle:
  - Probably varies a lot per kind of intersection, but in the example shown it looks like extrapolating a rectangle from opposite sides (for 4-way), for 3-ways can be a bit different but still a consistent breakdown of triangles
- Need to specify limits of range of angles for the roads to connect
- No input here around how the textures are generated, and I take it they are baking them to road-specific files, which is not really what we want to be doing. So this part remains to solve!

**Comment by TheDuckCow** [comment]
Thanks to Kitten on our team, here's at least one example of someone talking about how to approach procedural intersections. I think in the end we'll use a composite of techniques, but getting all the references we can is a solid start

https://twitter.com/noio_games/status/1704891239244382240

**Comment by TheDuckCow** [comment]
Another demonstration of an approach as well as user visual interaction: https://www.youtube.com/watch?v=3O897gbyiBk

This one is interesting, as the actual intersection is much more implied for the user, rather than it being an explicit object type. They seem to have a singular (equivalent) concept of RoadPoint. The icky party is that it means you end up creating multiple overlapping RoadPoints which still belong to each segments. But it's interesting it gives you some chance to wiggle-connect.

Two other videos I want to read more into: [intersection decals](https://www.youtube.com/watch?v=RF-NJNkZ0qE) and this [Houdini intersection solver](https://www.youtube.com/watch?v=-cjeq7GttH4). This last one gives some nice philosphy for how to think of the curves connecting the segments.

**Comment by TheDuckCow** [comment]
Another video I came across for Unreal for creating intersections using a more primitive approach, this one's interesting in a way because it's just making use of overlapping roads and then overriding the material at the end. Their process:

1. Create the road going through
2. Over-connect the roads to each other
3. Create an "override" material for the specific sections being added (hm, a concept like this could be nice for RoadPoints too)
4. The using decals they are then able to add lines back to the correct edges of the mesh

https://youtu.be/RCv3dFUjOf0?si=T9_Jh0wzFTuYDeqZ&t=760

I don't think the above is exactly what we'd want to do, but the approach of using layered materials to get the right edges could be interesting. It depends on a global detail texture being used, which is also something we would probably want to implement eventually anyways.

**Comment by TheDuckCow** [comment]
Another [example asset/system](https://blendermarket.com/products/next-street?ref=369), another [video](https://www.youtube.com/watch?v=6_478KOFEA0). See [here this](https://blendermarket.com/products/next-street?ref=369) time a blender addon called "Next Street" that does quite a lot.

- A vertex is our equivalent to a RoadPoint, edges are the actual streets
- The simplicity of creating intersections is hard to beat: extrude out from vertex from an existing point
- It works such that you can change the weight for how much it creases in or out of this point
- It has various preset profiles for the streets, e.g. with or without the sidewalk, barricades etc. I'm not sure if it has a way to do it for one edge at a time or if it's universal. A lot of the selling point here are all the assets it contains including: trees, benches (multiple kinds, lamppost, railings), trash cans; also with settings for the spacing of each of these too
  - This further more makes me think it would be helpful to define some kind of resource slot for RoadContainers which define the "style" of road, and this resource could itself contain arrays to elements and numbers of references. This way it's also easy to copy or even save to disk to quickly change the style of an entire city, while still giving you the option to change the style of single edges at a time.
- It DOES offer the ability to change the number of lanes, presets of 2, 4, 6 etc. And that is based on selecting single edges at a time. Not sure what it'd do for transitions between two lane counts without an intersection.
- Doe shave AI lanes for traffic, including through intersection

**Comment by TheDuckCow** [comment]
Something interesting I came across while poling through some Kitbash3D files, there's one of a T-intersection. What surprised me is that they are creating actual geometry for each material overlay, like lines and so forth (and in fact, the cross walk lines are actual geo which is raised up, but not just resting on the mesh, they are joined and cut)

This would be too much overhead for realtime gen I'd say so we still want to look at decal solutions, but interesting nonetheless.
<img width="1069" alt="Screen Shot 2024-09-20 at 11 55 41 AM" src="https://github.com/user-attachments/assets/01243222-bf8b-4795-b986-f780490b8b54">

**Comment by TheDuckCow** [comment]
As I keep coming across more references, here's yet another example of someone explaining and walking through how to do intersection creation in houdini, likely some points to learn in especially how they break the problem down:

https://www.youtube.com/watch?v=bTlY9ahThdQ

**Comment by TheDuckCow** [comment]
Looks like user spearhead1 over on reddit has made a pretty solid looking intersection generator a few months ago, reached out in hopes of collaboration or at least strategy sharing:

https://www.reddit.com/r/godot/comments/1dnrcki/procedural_geometry_road_builder_tool_bonus_godot/

**Comment by ChristopherHaws** [comment]
Not sure if this is useful or not, but I came accross these videos and javascript sample project which seem like they might be helpful:
Video: https://www.youtube.com/watch?v=V_C7L7zelz8&list=PLB0Tybl0UNfYoJE7ZwsBQoDIG4YN9ptyY&index=12
Demo: https://radufromfinland.com/projects/virtualworld/
Demo Source: https://github.com/gniziemazity/virtual-world

**Comment by Slluxx** [comment]
[Here is a unity example](https://www.youtube.com/watch?v=ZiHH_BvjoGk) i came across a while ago, which does explain everything very neatly. Depending on the information we can get from a path, this might be relatively easy to reproduce in godot.

**Comment by antonkurkin** [comment]
Here are my thoughts that may be useful in some way
I'm thinking that we can approach it from the perspective of AI lanes. I'm thinking that it's all can be done on a plane (y=0) for simplicity of operations and then "stretch" the result to the actual 3d geometry
basically if we can (automatically or manually - or both) create all the lane connections between road points we need - then we can chose every leftmost and rightmost lane that is connected to every the road point and extend it (left or right, for half a width and then do the shoulder). then we will intersect all the curves we get this way and here is our minimally-needed geometry
for example, here is a simple split with and without U-turn
![Image](https://github.com/user-attachments/assets/17b42257-e547-4f35-a630-3ecacae45389)
here are a couple more comples examples
![Image](https://github.com/user-attachments/assets/a0c53468-fc54-4b84-ac9e-8103b5946bcb)
I think it's good that it could be very visual and dynamic in editing - as well as using minimal space. as a downside there could be a lot of edge curves and while I think some of the already existing code for normal road generation could be used, I wouldn't expect that it will be very easy

**Comment by TheDuckCow** [comment]
FYI for anyone interested, we are holding a community discussion to kick off development for this feature, you're welcome to join here this coming Tuesday:

https://discord.gg/H4r6zS93?event=1418474791161303100

Thanks for helping jump in here @Picorims!

**Comment by TheDuckCow** [comment]
Based on our call, the following is proposed. We can debate and discuss this with the community from now till the end of this weekend (Sept 29th), at which point @Picorims would begin working on the feature. The proposed structure:

- RoadIntersections will be/remain as a class / node provided by the addon
  - They are exposed in the editor and are themselves Node3D's, but generally should not be manually placed by users
  - Instead, when converting a collection of RoadPoints into an intersection, it should automatically create the intersection object
  - In terms of scene hierarchy, intersection geometry can be a child of this intersection node directly
- The noteworthy change to debate: We will create a _new_ baseclass called [maybe `RoadEdge`? `RoadNode`? `RoadSlice`?]
  - RoadPoints would then be updated to extend from this new baseclass instead of Node3D directly
  - This class essentially has no functionality or code overhead. Used for typing primarily.
  - Then, the RoadIntersection actually extends from this common base instead of Node3D directly
  - This way, we can allow RoadPoints to connect to any other [RoadEdge/RoadNode/RoadSlice] so we can know which "direction" an intersection is located
- RoadContainers will get a new node reference which references a RoadIntersection
  - At most one RoadIntersection is relevant per RoadContainer; the RoadContainer overall should be essentially treated as a complete intersection (much like the provided prefab intersections are already today)
  - A single RoadContainer is either in the mode of a "normal road" (A->B) _or_ an intersection mode, and this is defined by the export var reference to the RoadIntersection
  - The intersection object is what defines the nature of how the intersection works, how geometry is created, etc.
- The RoadIntersection itself defines the origin point of the intersection, so you can move it around in 3D to offset it without having to move the RoadPoints themselves necessarily (but you can also still move the whole thing by moving the RoadContainer itself, too)
- RoadIntersections are siblings to RoadPoints in the same RoadContainer. All RoadPoints in this container would, at least as we expect, have one of their next/prior point references be the RoadIntersection, or be left blank because it is an "edge" for the RoadContainer.
  - To keep things strict and clean, we probably need to _require_ that every RoadPoint in the container is automatically connected to the Intersection object if it exists; tools like dissolving or adding new RoadPoints off of pre-existing ones should not work on an intersection container.
- The RoadIntersection defines a "generator" variable, which for now will always be a unique instance of the road_meshgen_ngon.gd resource
  - This resource file would contain all the logic for creating the mesh and UV itself based on input sets of RoadPoints and the intersection itself
  - By being a resource, we can define a variable interface in Godot depending on the type of intersection. ngons or maybe "vertex" intersections might have a rounding param to define how sharp the corners come together, while a roundabout might support the number of roundabout lanes.
  - We thought about making a baseclass resource script to extend from, but generally this should be created by the addon anyways so might not be necessary.

I want to follow up with more of a mockup demo of this process, as this really will define how convenient and natural the feature will be to use. But in essence, the UX flow as we're thinking of it so far:
1.  Start with a stretch of connected road, 3+ points.
2. By dragging another roadpoint off of another stretch of road, or with the Connector (+) tool enabled, overlap/drag towards to the middle roadpoint (or, maybe we allow "bisecting" in the middle of a long enough stretch). Visual indicator, maybe a circle or something, highlights _all_ the RoadPoints which would become part of the intersection
3. On pressing/releasing, the intersection is created - a rather big UI operation in terms of undo/redo and nodes being changed in the scene: 
   - a new container is generated
   - copies of the highlighted RoadPoints in question are duplicated into this new container
   - Each of these new edge RoadPoints are then marked as connected to their original points they are duplicated from (this is actually accomplished by updating export vars on the respective RoadContainers)
   - Then a new intersection object is added in the position of the original RoadPoint where the intersection was going to be created (if we have it snapping to a RoadPoint; if it can bisect a RoadSegment, then there's no actual reference RP, just the closest point along that segment)
   - Each internal roadpoint of this new container is then also marked as connected to this new intersection object
   - Automatically add an instance of the road_meshgen_ngon.gd resource to the intersection object

Not considered: all the rules around ensuring there's enough space, how to delete intersections (if you disconnect one side, does it get merged into the based RoadContainer?)

As discussed, here's the task breakdown we're thinking of:

- [ ] I'll crate a feature branch for this call 121-intersection-feature, based off of dev (and hopefully, the v0.8.0 release itself).
- [ ] @Picorims to work on the base structure layout (updating the intersection class, creating the extended "roadnode" idea, and the setup of the generator resource; in this step, we should alos define the connection of refreshing a RoadContainer -> refreshing/generation of geo for the intersection object
- [ ] @Picorims's Next step would be actually working on the road meshing for intersections, ngon scenario
- [ ] @TheDuckCow on UI tooling support, could be done in parallel to the actual meshing itself
- [ ] Thereafter, we can decide who's best equipped to look at UV support
- [ ] (merge the feature branch!)
- [ ] Some further distant future: a Roundabout generator class could be created as well.

If we can agree and align on this between now and the end of this coming weekend (Sept 29th?), then we'll be good to proceed.

**Comment by Picorims** [comment]
I would add the following:

- base class name could be `RoadGraphNode` to reuse graphs terminology and make it distinct from Godot's node terminology? Or `RoadVertex` (https://en.m.wikipedia.org/wiki/Graph_theory)
- The default resource should be made unique to avoid bad surprises upon editing them. But ideally when dragged from the UI it should be at user's discretion (not sure if the distinction can be done?) eventually there could be a force unique toggle somewhere (RoadContainer or RoadManager). But it can be a consideration for the future.
- I would not automatically destroy an intersection, but let the user do it if it wants, if there are exactly two `RoadPoint` in it. I would see a situation where you disconnect one road to connect another instead during game design (in which case you don't want all the tinkering go to trash unexpectedly). Or even for actual gameplay, unary roundabouts exists, for instance in front of some hotels. If that is controversial I'd also make a settings out of it (it is just a matter of automatically calling the function or doing so on button press).
- You can extend the intersection with additional roads by snapping to the intersection.
- You can shrink it by disconnecting an edge (the intersection's corresponding road point is deleted.
- Mesh refresh is triggered on edge/intersection position or count change, eventually throttled if expensive.
- The resource class could expose a method for querying the minimum retired distance from the intersection required to work properly. The container could then use this info to translates points too close along their facing direction vector. This would be extremely valuable in procedural generation, in which case it has to be resolved, manually or through the plug-in. In terms of UX it would also reduce manual tinkering needed. But it can be made as a separate issue for improvement. 
- UI interactions (at least the core subset) should have equivalent public methods to support them in procedural generation.

**Comment by TheDuckCow** [comment]
> - I would not automatically destroy an intersection, but let the user do it if it wants, if there are exactly two RoadPoint in it. I would see a situation where you disconnect one road to connect another instead during game design (in which case you don't want all the tinkering go to trash unexpectedly). Or even for actual gameplay, unary roundabouts exists, for instance in front of some hotels. If that is controversial I'd also make a settings out of it (it is just a matter of automatically calling the function or doing so on button press).
> - You can extend the intersection with additional roads by snapping to the intersection.
> - You can shrink it by disconnecting an edge (the intersection's corresponding road point is deleted.

On the "not auto delete" part: It's a good point. The thing I'm most "nervous" about our approach is that creating an intersection is a complicated process from a scenetree standpoint. Generally, I want to make sure that users don't have to care about RoadContainers / scene structure unless they really want to. If there's a lot of intersection creation and then removals, you end up with a bit of a mess of RoadContainers. But, we could also create some kind of merge/split tools to assist with this if needed I suppose.

I created this demo of what the UX could look like (with the magic of jumpcut editing, and re-using the pre-existing prefab intersections). The main difference to this example vs what we spoke about above, is that new RoadPoints are inserted much closer to the intersection, giving just enough space, rather than using the original edges. I think this makes more sense, since there could be vast distances between the roadpoints around the intersection point to begin with. You wouldn't realistically want the intersection to span the whole space (such as in these examples).

https://github.com/user-attachments/assets/8d18b9ad-8b80-4b9c-8f9c-85c8351c7a00


For your other points:

> base class name could be RoadGraphNode

Nice idea, I'm open to it

> default resource should be made unique

Yup agreed, we can make the tooling do this automatically, shouldn't be a problem

> Mesh refresh is triggered on edge/intersection position or count change, eventually throttled if expensive

I'd just update the mesh on release of the mouse when it comes to snapping in new RoadPoints. if you're instead just moving one RoadPoint "arm" of the intersection, which wouldn't change the intersection "shape" (from a topological standpoint), that could be interactive like the rest of the roads with some timer / low poly representation as you suggest.

> The resource class could expose a method for querying the minimum retired distance from the intersection required to work properly

I think this is a good idea. We'll have to think what it means to "push" RoadPoints out further if there's not enough space, or if we just create a warning and skip some steps. If we do "nudge" points further out, we need to be careful of chain reactions (e.g. the other end is already directly another intersection at its minimum space)

**Comment by Picorims** [comment]
I like the overall simplification and also think that the simpler it, the less risks we take and the easier it is to maintain. 

Regarding extra loops, I am afraid that it adds complexity to mesh generation if handled by the intersection. The intersection generator should only have to deal with a list of coordinates and not have to compute offsets itself especially when those are curves. So it must receive an offseted road point (which can be achieved with a public getter to get the computed offseted coordinates).

I don't think it drives away from organizing in containers, at least from what I understood with my own usage. Points within the same container can't work independently (said differently a container is a set of interdependent points). So you may want to cut them in multiple containers to load and unload regions. If you want multiple different textures, you need a new container everytime it changes. So it is kind of an atomic unit which size depends of the needs of the user. It just needs to be clear in their mind with visual indications and/or explanation in the doc. It could also be something to think about for LODs. In the current situation I don't see what other purpose it serves (again, still rather new to the plugin so there might be something I am not aware of as I don't use presets and the GUI myself).

Regarding not having to deal with merging, what happens if you want to form an intersection from two different containers then? Or you have to do it the other way around and create a branch of a road? I find the later a bit less intuitive, even if it is not a big deal either (you could support both with magic in the background).

It still needs spec regarding what can connect to what. Can an intersection node be a container edge? (I'd say no). Can two intersection nodes be connected? (I'd say no for now, could be interesting in the future for pilule roundabouts or merged intersections but way too much complexity and burden added for a first iteration).

Otherwise yes I think it will be much more robust this way.

**Comment by Picorims** [comment]
@TheDuckCow 
> Then during geometry generation, the intersection "knows" it only extends 10m in the direction of RP10, so it creates geo that far in that direction. It calculates the Transform3D at a curve offset of 10m

This is what worries me when I say I am afraid that it adds complexity. The intersection get to take into account things that it should not have to care about in my opinion. The generator should just have to know explicitly to which points it extends with what orientation (if prompted from the road point, it is free to return some virtual coordinates instead of just reading its global position).

> The alternative is, to avoid excessively large intersections, that creating an intersection likely requires inserting another RoadPoint between the initial RP and the intersection, to serve as a good the boundary point.

I think this is easier and better if both points are far away from each other. You do not get a very stretched intersection, but still give an easy way to fine-tune it: moving and rotating the newly added road point (or existing one if considered close enough eventually).

> Question: Do RoadPoints have magnitude handlebars that extend into the intersection? Or do intersections always linearly/statically eased reach out towards the given RoadPoints? What if the connecting RoadPoints are twisted relative to the intersection? This has some implications.

I'd say that the road point itself (position and orientation) is the control point. It could be interesting to have finer grained control over the edge rounding, though if considered I'd keep it stored on the road intersection side in some sort of array or map to keep things decoupled. It would be nitpicking though I think, and might better be put in a later improvement if considered?

Twisting is not really an issue I think, it just moves the road edges points, which in this case *have* to be computed for mesh generation no matter what. I think it probably even simplifies the computation of their locations by not having to figure out the vector direction before computing edge locations.

> Maybe under the hood we actually just generate curves for each out the outermost RoadLanes on all sides, and use that as a guide for the outside shape?

It'd be ideal but does not sound trivial at all. The only approach that comes to mind is doing it the shader way, creating a plane in which each curve is an SDF, computing the union, and then somehow retrieving its border as baked curves. Or figuring out all curve intersections and computing the largest polygon encapsulating it but that would still create wasted space in some scenarios. I am not closed to it but it would need deeper thinking before diving into code I think. At least when getting to the mesh generation step at least.

In the sketch I provided it remains possible having the circle be of radius max(road_point_lanes_count for each road point). Which admittedly can go into wasting a bunch of space very easily if the gap between points is large.

Another thing coming to mind, I would not require agent lanes to exist for mesh generation if someone do not need them (default to all possible curves?).

---

OK with the rest.

**Comment by TheDuckCow** [comment]
Thanks for chiming in @ToasterBiome - this will definitely make working with intersections easier, although we also want to make sure working with RoadContainers is fluid as well. On that topic, would you say your issues with RoadContainer intersections is more to do with bugs around the connection tool not letting you complete connections, or more with issues around trying to move intersections after the fact?

**Comment by Picorims** [comment]
> I would argue that could actually belong to the RoadPoint instead, or at least as an override to a value coming from the intersection.

I like the override idea. And I'd actually think more intuitive. Some reasons I didn't stick to it when thinking through it first is:
- is the property visible on points not connected to an intersection?
- If so how to indicate that it has no effect? Or how to hide it?
- Do you keep the value if disconnected from its intersection?
- If it links two intersections, can you have a different rounding for each?

But yeah it can be envisioned later, and it might actually help to see how the architecture comes out and what limitations it gives to make decisions on that. I don't expect much friction from adding it later on.

The road lane part is a fair point.

I can leave in more time if needed, it is not like it is urgent on my side and I still have 2 good months in front of me.

**Comment by TheDuckCow** [comment]
> - is the property visible on points not connected to an intersection?
> - If so how to indicate that it has no effect? Or how to hide it?
> - Do you keep the value if disconnected from its intersection?

I think all of these can be waved away if we consider the property "internal data" and put it under that internal data section. Eventually the goal would be to actually put those properties into an `export_storage` group ([reference](https://docs.godotengine.org/en/4.4/tutorials/scripting/gdscript/gdscript_exports.html#export-storage)), which would mean it won't be visible in the UI. Then, it could be just visualized through a combination of 3D gizmo as well as more dynamic drawing in the RoadPoint inspector panel (the top section which has the lane +/- buttons and copy/paste buttons).

> If it links two intersections, can you have a different rounding for each?

... this however, is a good counterpoint! I suppose we could solve that issue with a tiny bit more complexity by having both forward and reverse rounding properties, but tbh it's probably not worth the extra complexity.

The other thing to consider is how this translates to other kinds of intersections we might make, like on/offramps (which should not really influence the "primary" road shape) and roundabouts. I suppose if we just call it "weight" or something like that, each intersection generator could decide if/how to leverage this weight value?

> I can leave in more time if needed, it is not like it is urgent on my side and I still have 2 good months in front of me.

Again, all of this is not necessary for v1.0 of intersection support, but the takeaway is there'll be a way to make it work as long as we have the end goal in mind. And you don't need to sign yourself up for 100% of the work :)

Trying to wrap up the v0.8.0 release today, then I'll take a stab at the geometry updates.

**Comment by Picorims** [comment]
> > - is the property visible on points not connected to an intersection?
> > - If so how to indicate that it has no effect? Or how to hide it?
> > - Do you keep the value if disconnected from its intersection?
> 
> I think all of these can be waved away if we consider the property "internal data" and put it under that internal data section. Eventually the goal would be to actually put those properties into an `export_storage` group ([reference](https://docs.godotengine.org/en/4.4/tutorials/scripting/gdscript/gdscript_exports.html#export-storage)), which would mean it won't be visible in the UI. Then, it could be just visualized through a combination of 3D gizmo as well as more dynamic drawing in the RoadPoint inspector panel (the top section which has the lane +/- buttons and copy/paste buttons).
> 
> > If it links two intersections, can you have a different rounding for each?
> 
> ... this however, is a good counterpoint! I suppose we could solve that issue with a tiny bit more complexity by having both forward and reverse rounding properties, but tbh it's probably not worth the extra complexity.
> 
> The other thing to consider is how this translates to other kinds of intersections we might make, like on/offramps (which should not really influence the "primary" road shape) and roundabouts. I suppose if we just call it "weight" or something like that, each intersection generator could decide if/how to leverage this weight value?
> 
> > I can leave in more time if needed, it is not like it is urgent on my side and I still have 2 good months in front of me.
> 
> Again, all of this is not necessary for v1.0 of intersection support, but the takeaway is there'll be a way to make it work as long as we have the end goal in mind. And you don't need to sign yourself up for 100% of the work :)
> 
> Trying to wrap up the v0.8.0 release today, then I'll take a stab at the geometry updates.

If made internal I would just at least expose a public API to change it programmatically. 

No hard feelings on having one or two values. One is probably easier.

I like the weight idea. For roundabouts or would influence the branching width and length. For an offramp, it could control the length of it along side the other road to make some (de)insertion line (very common in France).

But yeah not for v1 for sure. As for how far I'll go beyond it'll depends on my own needs and how far I can go on my game before being back to a full-time job. Variants could still be probably beneficial to me though. Also if the foundations are there it probably make it easier for others to contribute afterwards as well.

**Comment by TheDuckCow** [comment]
Follow up on the topology points here - I recorded this video as I explored how to create the actual geo inside of Blender. While I know we talked about you not worrying about UVs yet, I do want to make sure the topology solution we _do_ come up with at least will somewhat work for UV based textures after. For that reason, a triangle fan/circle I've concluded would be too limiting. At the end of the video, I proposed a "simple" greedy algorithm for making a grid shape. But yes there's a lot of complexity even with that, as you have to figure out the "corners", though maybe we can know that from the vertex indices anyways.

Of course, we could always start with the triangle fan to just get the feature out the door. Maybe if time is limited, you can begin with that, but it would be good to have that e.g. in its own function or otherwise parsed out so we can change out the "filling" approach later to be more UV friendly.

https://youtu.be/jhXS9oACfjI

And here's the blend file that has the example meshes, so you could play around with that.

[intersection_geo_v04.blend.zip](https://github.com/user-attachments/files/22706137/intersection_geo_v04.blend.zip)

The blend file here addresses examples of 3, 4, and 5 way intersections. I have it in my head it would generalize to 6 or more, but I guess limiting to 5 is a decent starting place for manually trying it out. It also describes an approach for out to define the continuous, outside lanes to each corresponding neighbor RoadPoints along the outside, and the shoulder shapes. By defining the outside shapes like this, we then can know the outside lanes follow the same shape as they would if that were an ordinary stretch of RoadSegment, meaning the placement of RoadLane connections there will follow nicely too, which is important to maintain. So start by building the shoulder + matched up outside lane(s) for each arc between RoadPoints is the way to go I think.

Once you have a watch, I'm curious if you have any further thoughts. 

Screenshot of the 5-way intersection sample below, with a not-so-expertly placed UV grid but at leas illustrates approximately what we could get while maintaining the use of the same trimsheet texture material with the rest of the roads (scale is off, but still illustrates it can be placed better than a full triangle fan). The highlighted quads are the "matching lanes" between adjacent RoadPoints, ie the ones that would essentially be identical to the ones created normal RoadSegments were the given pair of RoadPoints connected together directly.

<img width="503" height="457" alt="Image" src="https://github.com/user-attachments/assets/d75b827a-cf19-47b0-8516-fbc16b653098" />

By comparison, the triangle fan approach, which I think of as more similar to your original proposal (and again, we could still use this as an interim solution to get a version 1 ready)

<img width="522" height="424" alt="Image" src="https://github.com/user-attachments/assets/0a2749d8-e5e4-4e80-8115-c4f99dbfb118" />

**Comment by Picorims** [comment]
@TheDuckCow 

Thoughts while watching the video:
- For the grid algorithm, I think I'll have to just try a bunch of stuff, but some ideas include:
    - To detect the angles, it could be by knowing the edge circumference border or whatever we call it, finding the n vertices with the smallest acute angle (n being the number of road edges), looping through the whole border.
    - A very naive approach could be to compute an axis aligned bounding box enclosing all the border vertices, use a 2D loop to setup a grid of vertices which only adds one if inside the border (Godot has a 2D algorithm for that), then attaching it to the border. The center will look clean, the edges however will look very weird.
    - An algorithm that, knowing corners and border vertices in between, do optimization until reaching an ideal square case from which we build a grid with proportional distances along the axes. But the optimizations would have to be UV friendly themselves.
        - It could be also about cutting it in N squares.
- Regarding the missing lane case at 8:00, we should know at this point that there is a gap. We can use the distance between vertices of both borders so that below a given distance, we create quads between them. If the length is big, we can have multiple columns of quads. Which mean it can't be an arbitrary value but based onto a parameter (number of missing lanes times a factor slightly above one?)
- At 11:21, using the curve approach wouldn't shock me as in France for instance in such situation the pinch is paint on the road and the actual asphalt border is a curve. But otherwise, I would say simultaneously build "as usual" the shoulder of both side until some distance threshold at which point they join themselves (eventually using the tangent vector of last edges to figure out the optimal location for the joint vertex, not sure how tricky this one would be, probably solvable in 2D using trigonometry). Which save the main lanes texturing until that point. But then I would expose an option to switch between curve and pinch, because you can't guess what the end user wants (this layout could also very well be a city intersection expecting a curve). Then the inside would be the same as other cases. In terms of road paint it should look somewhat OK.
- 13:40, agreed it is not a good solution in the end. And it was also a triplanar based solution so if going with quads it doesn't work.
- 18:00, is also a solution. But that's why I'd see if there isn't some way to reach an optimal "square" or "rectangle" of vertices to minimize the risk of stretches.

For the center approach, that's why I wanted to create intermediate ones with smaller decremental radiuses, but you inevitably get distortion. It was just because it is the most trivial to do, but using noise texturing myself it will look very odd and off.

---

Regarding the architecture, I'll probably make a summary comment of all the spec that came out of the thread to have a clearer view of the whole thing before getting into code. Basically a bullet list/tree of all requirements to have some acceptance criteria on which we can perform tests and ensure things work correctly. And to be sure nothing is missed in the process (or as little as possible at least).

**Comment by Picorims** [comment]
# Spec summary (cc @TheDuckCow ):

## Classes
- `RoadIntersection` (point):
    - kept
    - exposed in editor, Node3D, but not for user
    - collection of road points as input of a new intersection object (point).
    - instanced by converting a `RoadPoint` or bisecting a segment made of two `RoadPoint`s.
        - Does **NOT** create a new road container.
    - intersection node has geometry as child.
    - holds specs (kind, params, etc.) through a resource (for now `road_meshgen_ngon.gd`.
        - has mesh/UV generation logic, takes info as input, returns mesh as output.
        - The resource type define the intersection type and its params (which can then be different)
        - baseclass resource?
        - The default one must be made unique.
        - exposes a method to know the minimum required distance from the intersection (push a warning if condition not met at mesh gen and don't generate anything). The owner should check first and relocate `RoadPoint`s if needed, if they are **NOT** a container edge.
        - The owner should insert a `RoadPoint` between an intersection point and another `RoadPoint`, if considered too far (threshold to be defined based on road lanes count).
    - Is the intersection center/origin.
    - Moving edges is done through linked road points.
    - sibling to road points in a `RoadContainer` (one container can have multiple, one, or no intersection(s)).
    - each `RoadPoint` is connected to the intersection only, the other side being a road edge.
- baseclass `RoadGraphNode`
    - from which `RoadPoint` and `IntersectionPoint` extend.
    - no logic, for types only.
    - can connect to any `RoadGraphNode`, direction should be known.
    - If deleted, it creates a hole.
- `RoadContainer`
    - can have as many road points and intersection points as needed.
    - normal or intersection mode.
- `RoadPoint`
    - now only generate a full mesh if the next sibling/successor is **NOT** an intersection, otherwise goes until the threshold of the generator is reached.
    - > (right now we use a greedy loading approach where the first roadpoint to "get" to a segment is the one to trigger generating its geometry)

# UX / User story

> 1. Start with a stretch of connected road, 3+ points.
> 2. By dragging another roadpoint off of another stretch of road, or with the Connector (+) tool enabled, overlap/drag towards to the middle roadpoint (or, maybe we allow "bisecting" in the middle of a long enough stretch). Visual indicator, maybe a circle or something, highlights all the RoadPoints which would become part of the intersection
> 3. On pressing/releasing, the intersection is created - a rather big UI operation in terms of undo/redo and nodes being changed in the scene: 
>     - a new container is generated
>     - copies of the highlighted RoadPoints in question are duplicated into this new container
>     - Each of these new edge RoadPoints are then marked as connected to their original points they are duplicated from (this is actually accomplished by updating export vars on the respective RoadContainers)
>     - Then a new intersection object is added in the position of the original RoadPoint where the intersection was going to be created (if we have it snapping to a RoadPoint; if it can bisect a RoadSegment, then there's no actual reference RP, just the closest point along that segment)
>     - Each internal roadpoint of this new container is then also marked as connected to this new intersection object
>     - Automatically add an instance of the road_meshgen_ngon.gd resource to the intersection object

In addition:
- You can extend the intersection with additional roads by snapping to the intersection.
- You can shrink it by disconnecting an edge (the intersection's corresponding road point is deleted).
- new RoadPoints are inserted much closer to the intersection.
- mesh is refreshed on mouse release in a UI setting, otherwise through a method call in a procedural setting.
- mesh refreshed as low poly (throttled) if just moving points.

![video user story](https://private-user-images.githubusercontent.com/2958461/493719743-8d18b9ad-8b80-4b9c-8f9c-85c8351c7a00.mp4?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTk3ODIzNTQsIm5iZiI6MTc1OTc4MjA1NCwicGF0aCI6Ii8yOTU4NDYxLzQ5MzcxOTc0My04ZDE4YjlhZC04YjgwLTRiOWMtOGY5Yy04NWM4MzUxYzdhMDAubXA0P1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MTAwNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTEwMDZUMjAyMDU0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NWRjOTM5NjAxNmY0ZTIwZTNmNjkzZDNlMzFlMDY1OTA4ZjNlMDkxMGE3NTZmNjRiYTA4MzJlOTU1ZWUyYzYzMyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.ApODKl8VWdCi5ba991gsZNwo1igkh0OKGD1SPKA0I9w)

# pre-made intersections

They should still be supported while not interfering with the new architecture.

# Mesh generation

Only the N-gon case is considered.
Requirements:

# Out of scope
- dead ends (depends of how it evolves and if it is trivial or require a dedicated ticket?)
- tools for "moving" a RoadPoint into one container or another, as well as merging or splitting RoadContainers.

**Comment by Picorims** [comment]
(accidentally sent, not ready)

**Comment by TheDuckCow** [comment]
Quite the writeup indeed! And definitely appreciated, speaking as a PM myself :) This has the OK from my standpoint to begin ✅ just some additional details below:

---

- I've been thinking for a little bit about setting up a (public) github project for tracing community and release initiatives. If you'd find that easier to have a kanban style of priority order tasks and progress columns, just let me know - it's pretty quick to set up.
- In terms of creating sub tasks, feel free to go right ahead. I imagine you'd want to break up point 2 under Steps into further subtasks. Even if you do use the new "parent" feature, I presume the child issues end up as standalone issues in the overall issues tab right? In which case, I'd just request using a common prefix epci/feature name. e.g. `ProcInt:` such that any given issue is then titled `ProcInt: Implement RoadGraph structure` as an example.
- Branch now created: https://github.com/TheDuckCow/godot-road-generator/tree/feature-121-proc-intersections
- Calling out: your video and private image upload both aren't resolving for me, the CDN links (which is what I believe those are) aren't the same as the static user content links. maybe try dragging and dropping the source files back in? or if they are just referencing the videos/content further up in the thread history, feel free to just link to those comments instead.
- Once that's good, I'll probably edit the root issue comment with the content so it's easy to get back to, and we can host the subtask links there as well.
- Finally, writeup overall looks good - just one point:
  > each RoadPoint is connected to the intersection only, the other side being a road edge ;
  
  Just to make sure we're on the same page, since RoadPoints and RoadIntersections are siblings now, I think there should be no problem with having a single RoadPoint be connected to two simultaneous RoadIntersections, ie on both sides.

**Comment by Picorims** [comment]
Made the fix here:
> Cannot have an intersection point as edge. In this scenario, at least one RoadPoint must be between said RoadIntersections.

I think your quote is not relevant anymore as this is from before it was decided that the intersection is not its own container anymore (I went through the whole thread and edited my draft as information changed).

Copied the image in the comment and linked the comment to the video.

**Comment by TheDuckCow** [comment]
Great, sounds good regarding sub issues. Looks like you're all good to proceed - chime in here as you run into anything and I'll try to respond promptly! If it's helpful, we could also proactively schedule a checkin time (in case it's easier to live chat through a number of little questions that inevitably come up). Just let me know. I'm fairly free Friday morning as it stands right now, which would be your afternoon.

**Comment by Picorims** [comment]
This works for me. I had that in mind under different names (edge terminology) but I like "branch" more. Not sure what you mean by automatically reacting to props and translation (maybe some Godot aspect I don't know of especially if it is something else than signals). But I'll start with these methods. Which is an opportunity to test procedural generation in a separate scene and see how it reacts.

**Comment by TheDuckCow** [comment]
Updating this thread to say we've got some good progress here - though still a ways to go (and hmm we seem to have some extra global coordinates leaking in somewhere)

https://github.com/user-attachments/assets/f560bcff-69dc-472d-ba6f-4467c0b0ac52
